controllers/networkfence_controller.go

```diff
@@ -39,17 +67,79 @@ type NetworkFenceReconciler struct {
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
 func (r *NetworkFenceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// fetch NetworkFence object instance
+	nwFence := &csiaddonsv1alpha1.NetworkFence{}
+	err := r.Get(ctx, req.NamespacedName, nwFence)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("NetworkFence resource not found or deleted")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// validate NetworkFence object so as its parameters are neither empty nor nil.
+	err = validateNetworkFenceSpec(nwFence)
+	if err != nil {
+		logger.Error(err, "failed to validate NetworkFence spec")
+
+		nwFence.Status.Result = csiaddonsv1alpha1.FencingOperationResultFailed
+		nwFence.Status.Message = fmt.Sprintf("Failed to validate Networkfence parameters: %v", util.GetErrorMessage(err))
+		statusErr := r.Client.Status().Update(ctx, nwFence)
+		if statusErr != nil {
+			logger.Error(statusErr, "Failed to update networkfence status")
+
+			return ctrl.Result{}, statusErr
+		}
+
+		// invalid parameter, do not requeue
+		return ctrl.Result{}, nil
+	}
+
+	logger = logger.WithValues("DriverName", nwFence.Spec.Driver, "CIDRs", nwFence.Spec.Cidrs)
 
-	// TODO(user): your logic here
+	client, err := r.getNetworkFenceClient(nwFence.Spec.Driver, "")
+	if err != nil {
+		logger.Error(err, "Failed to get NetworkFenceClient")
+		return reconcile.Result{}, err
+	}
+
+	// check if the networkfence object is getting deleted, if yes, then send
+	// UnfenceCluster request to the driver.
+	// After the UnfenceCluster request is processed, the finalizer will be removed
+	// so that the networkfence object can be deleted gracefully.
+	if !nwFence.GetDeletionTimestamp().IsZero() {
+		if util.ContainsInSlice(nwFence.GetFinalizers(), networkFenceFinalizer) {
+
+			err := r.unfenceClusterNetwork(ctx, nwFence, client, logger)
+			if err != nil {
+				logger.Error(err, "failed to unfence cluster network")
+				return ctrl.Result{}, err
+			}
+		}
+		logger.Info("NetworkFence object is terminated, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	// add finalizer to the networkfence object if not already present.
+	if err := r.addFinalizerToNetworkFence(ctx, &logger, nwFence); err != nil {
+		logger.Error(err, "Failed to add NetworkFence finalizer")
+		return reconcile.Result{}, err
+	}
+
+	err = r.fenceClusterNetwork(ctx, nwFence, client, logger)
+	if err != nil {
+		logger.Error(err, "failed to fence cluster network")
+		return ctrl.Result{}, err
+	}
+	nwFence.Status.Result = csiaddonsv1alpha1.FencingOperationResultSucceeded
```

Comment:
Add a success message too. 
Otherwise it maybe stuck with a failure message ?

Reply:
Good suggestion, updated